<a href="https://colab.research.google.com/github/soniacgm/Deep_Learning/blob/cnn/Tarea_v03.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Dado que el entrenamiento de redes neuronales es una tarea  muy costosa, **se recomienda ejecutar el notebooks en [Google Colab](https://colab.research.google.com)**, por supuesto también se puede ejecutar en local.

Al entrar en [Google Colab](https://colab.research.google.com) bastará con hacer click en `upload` y subir este notebook. No olvide luego descargarlo en `File->Download .ipynb`

**El examen deberá ser entregado con las celdas ejecutadas, si alguna celda no está ejecutadas no se contará.**

El examen se divide en tres partes, con la puntuación que se indica a continuación. La puntuación máxima será 10.

- [Actividad 1: Redes Densas](#actividad_1): 5.5 pts
    - Correcta normalización: máximo de 0.5 pts
    - [Cuestión 1](#1.1): 1 pt
    - [Cuestión 2](#1.2): 1 pt
    - [Cuestión 3](#1.3): 0.5 pts
    - [Cuestión 4](#1.4): 0.5 pts
    - [Cuestión 5](#1.5): 0.5 pts
    - [Cuestión 6](#1.6): 0.5 pts
    - [Cuestión 7](#1.7): 0.5 pts
    - [Cuestión 8](#1.8): 0.5 pts


- [Actividad 2: Redes Convolucionales](#actividad_2): 4.5 pts
    - [Cuestión 1](#2.1): 1 pt
    - [Cuestión 2](#2.2): 1.5 pt
    - [Cuestión 3](#2.3): 0.5 pts
    - [Cuestión 4](#2.4): 0.5 pts
    - [Cuestión 5](#2.5): 0.5 pts
    - [Cuestión 6](#2.6): 0.5 pts


In [ ]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np

<a name='actividad_1'></a>
# Actividad 1: Redes Densas

Para esta primera actividad vamos a utilizar el [boston housing dataset](https://www.cs.toronto.edu/~delve/data/boston/bostonDetail.html). Con el que trataremos de predecir el precio de una casa con 13 features.

**Puntuación**: 

Normalizar las features correctamente (x_train, x_test): 0.5 pts 

- Correcta normalización: máximo de 0.5 pts
- [Cuestión 1](#1.1): 1 pt
- [Cuestión 2](#1.2): 1 pt
- [Cuestión 3](#1.3): 0.5 pts
- [Cuestión 4](#1.4): 0.5 pts
- [Cuestión 5](#1.5): 0.5 pts
- [Cuestión 6](#1.6): 0.5 pts
- [Cuestión 7](#1.7): 0.5 pts
- [Cuestión 8](#1.8): 0.5 pts



In [ ]:
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.boston_housing.load_data(
    path='boston_housing.npz',
    test_split=0.2,
)
print('x_train, y_train shapes:', x_train.shape, y_train.shape)
print('x_test, y_test shapes:', x_train.shape, y_train.shape)
print('Some prices: ', y_train[:5])

Normalizamos los datos porque los valores de las variables tienen escala diferente. En cada caso restamos la media y dividimos por la desviación estándar

In [ ]:
train_mean = np.mean(x_train, axis=0)
train_std = np.std(x_train, axis=0)
x_train = (x_train - train_mean) / train_std

train_mean = np.mean(x_test, axis=0)
train_std = np.std(x_test, axis=0)
x_test = (x_test - train_mean) / train_std




<a name='1.1'></a>
## Cuestión 1: Cree un modelo secuencial que contenga 4 capas ocultas(hidden layers), con más de 60 neuronas  por capa, sin regularización y obtenga los resultados.

Puntuación: 
- Obtener el modelo correcto: 0.8 pts
- Compilar el modelo: 0.1pts
- Acertar con la función de pérdida: 0.1 pts

In [ ]:
# Fijamos una semilla para asegurar que el resultado es reproducible

tf.random.set_seed(1)


In [ ]:
# Vamos a diseñar la red a través de una API funcional porque el modelo es complejo

# definimos la entrada
inputs = keras.Input(shape=(13,), name='input_layer')

# Ahora definimos las capas ocultas que constituyen la red y la relación entre ellas. Cada capa incluye más de 60 neuronas. Utilizamos la función de 
# activacion relu porque es la más adecuada para este tipo de problema de tipo regresión

layer_1 = layers.Dense (64, activation='relu',
                        name = 'layer_1') (inputs)
layer_2 = layers.Dense (64, activation='relu',
                        name = 'layer_2') (layer_1)
layer_3 = layers.Dense (64, activation='relu',
                        name = 'layer_3') (layer_2)
layer_4 = layers.Dense (64, activation='relu',
                        name = 'layer_4') (layer_3)
# al final definimos la capa de salida                        
outputs = layers.Dense (1, name = 'output_layer') (layer_4) 
# y creamos el modelo 
model = keras.Model(inputs=inputs, outputs=outputs, name='my_model') 


In [ ]:
model.summary()

Model: "my_model"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_layer (InputLayer)     [(None, 13)]              0         
_________________________________________________________________
layer_1 (Dense)              (None, 64)                896       
_________________________________________________________________
layer_2 (Dense)              (None, 64)                4160      
_________________________________________________________________
layer_3 (Dense)              (None, 64)                4160      
_________________________________________________________________
layer_4 (Dense)              (None, 64)                4160      
_________________________________________________________________
output_layer (Dense)         (None, 1)                 65        
Total params: 13,441
Trainable params: 13,441
Non-trainable params: 0
______________________________________________________

In [ ]:
# Compilación del modelo
# Como se trata de un problema de regresión utilizamos el MAE
model.compile(optimizer='adam', 
                  loss='mse',
                  metrics=['mae'])

In [ ]:
# No modifique el código
model.fit(x_train,
          y_train,
          epochs=200,
          batch_size=32,
          validation_split=0.2,
          verbose=1)

In [ ]:
# No modifique el código
results = model.evaluate(x_test, y_test, verbose=1)
print('Test Loss: {}'.format(results))

4/4 [==============================] - 0s 3ms/step - loss: 18.9684 - mae: 2.7377
Test Loss: [18.96837615966797, 2.737718105316162]


<a name='1.2'></a>
## Cuestión 2: Utilice el mismo modelo de la cuestión anterior pero añadiendo al menos dos técnicas distinas de regularización.

Ejemplos de regularización: [Prevent_Overfitting.ipynb](https://github.com/ezponda/intro_deep_learning/blob/main/class/Fundamentals/Prevent_Overfitting.ipynb)

Puntuación:

- Obtener el modelo con la regularización: 0.8 pts
- Obtener un `test loss` inferior al anterior: 0.2 pts


**REGULARIZACIÓN TIPO 1. REDUCIR EL BATCH_SIZE**

In [ ]:
# Fijamos una semilla para asegurar que el resultado es reproducible

tf.random.set_seed(1)


In [ ]:
# Creamos el modelo

model = tf.keras.models.Sequential()
inputs = keras.Input(shape=(13,), name='input_layer')

layer_1 = layers.Dense (64, activation='relu', name = 'layer_1') (inputs)
layer_2 = layers.Dense (64, activation='relu', name = 'layer_2') (layer_1)
layer_3 = layers.Dense (64, activation='relu', name = 'layer_3') (layer_2)
layer_4 = layers.Dense (64, activation='relu', name = 'layer_4') (layer_3)
                       
outputs = layers.Dense (1, name = 'output_layer') (layer_4) 

model = keras.Model(inputs=inputs, outputs=outputs, name='my_model') 

In [ ]:
# Compilamos
model.compile(optimizer='adam', 
                  loss='mse',
                  metrics=['mae'])

In [ ]:
# Establecemos un batch_size de 16 para ver qué impacto tiene una reducción de este parámetro como método de regulación, que en el ejercicio anterior era de 32

batch_size=16

In [ ]:
# No modifique el código
model.fit(x_train,
          y_train,
          epochs=200,
          batch_size=batch_size,
          validation_split=0.2,
          verbose=1)

Epoch 1/200
21/21 [==============================] - 1s 12ms/step - loss: 540.8517 - mae: 21.4170 - val_loss: 523.5873 - val_mae: 20.8720
Epoch 2/200
21/21 [==============================] - 0s 4ms/step - loss: 377.4734 - mae: 17.1820 - val_loss: 126.5977 - val_mae: 8.2211
Epoch 3/200
21/21 [==============================] - 0s 4ms/step - loss: 87.8150 - mae: 7.1112 - val_loss: 55.2201 - val_mae: 5.3547
Epoch 4/200
21/21 [==============================] - 0s 5ms/step - loss: 41.0133 - mae: 4.5732 - val_loss: 28.8950 - val_mae: 3.9603
Epoch 5/200
21/21 [==============================] - 0s 4ms/step - loss: 20.4728 - mae: 3.1966 - val_loss: 22.3934 - val_mae: 3.6285
Epoch 6/200
21/21 [==============================] - 0s 4ms/step - loss: 20.7162 - mae: 3.1828 - val_loss: 18.8070 - val_mae: 3.2651
Epoch 7/200
21/21 [==============================] - 0s 4ms/step - loss: 15.2096 - mae: 2.8245 - val_loss: 17.6060 - val_mae: 3.2310
Epoch 8/200
21/21 [==============================] - 0s 4ms/s

In [ ]:
# No modifique el código
results = model.evaluate(x_test, y_test, verbose=1)
print('Test Loss: {}'.format(results))

4/4 [==============================] - 0s 3ms/step - loss: 14.6281 - mae: 2.7003
Test Loss: [14.628083229064941, 2.7002830505371094]


**REGULARIZACIÓN TIPO 2. REDUCCIÓN A UN MODELO MÁS SIMPLE**

In [ ]:
# Fijamos una semilla para asegurar que el resultado es reproducible

tf.random.set_seed(1)


In [ ]:
# definimos la entrada
inputs = keras.Input(shape=(13,), name='input_layer')

# Utilizamos la reducción a un modelo más simple como técnica de regularización. Para ello nos quedamos con solo dos capas

layer_1 = layers.Dense (64, activation='relu', name = 'layer_1') (inputs)

layer_2 = layers.Dense (64, activation='relu', name = 'layer_2') (layer_1)

# al final definimos la capa de salida                        
outputs = layers.Dense (1, name = 'output_layer') (layer_1) 

# y creamos el modelo 
model = keras.Model(inputs=inputs, outputs=outputs, name='my_simpler_model')

In [ ]:
# Compilación del modelo
model.compile(optimizer='adam', 
                  loss='mse',
                  metrics=['mae'])

In [ ]:
#batch_size=16

In [ ]:
# No modifique el código
model.fit(x_train,
          y_train,
          epochs=200,
          batch_size=batch_size,
          validation_split=0.2,
          verbose=1)

Epoch 1/200
21/21 [==============================] - 1s 10ms/step - loss: 550.1998 - mae: 21.6895 - val_loss: 608.6171 - val_mae: 22.9587
Epoch 2/200
21/21 [==============================] - 0s 3ms/step - loss: 531.2937 - mae: 21.3235 - val_loss: 577.7939 - val_mae: 22.2890
Epoch 3/200
21/21 [==============================] - 0s 4ms/step - loss: 482.6608 - mae: 19.9550 - val_loss: 542.3962 - val_mae: 21.5047
Epoch 4/200
21/21 [==============================] - 0s 3ms/step - loss: 492.6026 - mae: 20.1568 - val_loss: 499.1669 - val_mae: 20.5243
Epoch 5/200
21/21 [==============================] - 0s 4ms/step - loss: 416.3467 - mae: 18.6559 - val_loss: 450.1573 - val_mae: 19.3493
Epoch 6/200
21/21 [==============================] - 0s 4ms/step - loss: 377.3060 - mae: 17.3141 - val_loss: 395.2890 - val_mae: 17.9299
Epoch 7/200
21/21 [==============================] - 0s 5ms/step - loss: 316.8962 - mae: 15.9372 - val_loss: 340.6300 - val_mae: 16.3957
Epoch 8/200
21/21 [=====================

In [ ]:
# No modifique el código
results = model.evaluate(x_test, y_test, verbose=1)
print('Test Loss: {}'.format(results))

4/4 [==============================] - 0s 3ms/step - loss: 18.6685 - mae: 2.8164
Test Loss: [18.66849708557129, 2.816351890563965]


**REGULARIZACIÓN TIPO 3. DROPOUT** 

In [ ]:
# Fijamos una semilla para asegurar que el resultado es reproducible

tf.random.set_seed(1)


In [ ]:
model = tf.keras.models.Sequential()

# definimos la entrada
inputs = keras.Input(shape=(13,), name='input_layer')

# Añadimos el dropout en cada  capa

# capa oculta 1
layer_1 = layers.Dense (64, activation='relu', name = 'layer_1') (inputs)
layer_1 = layers.Dropout (0.5, name= 'dropout_layer1')(layer_1)

# capa oculta 2
layer_2 = layers.Dense (64, activation='relu', name = 'layer_2') (layer_1)
layer_2 = layers.Dropout (0.5, name= 'dropout_layer2')(layer_2)

# capa oculta 3
layer_3 = layers.Dense (64, activation='relu', name = 'layer_3') (layer_2)
layer_3 = layers.Dropout (0.5, name= 'dropout_layer3')(layer_3)

# capa oculta 4
layer_4 = layers.Dense (64, activation='relu', name = 'layer_4') (layer_3)
layer_4 = layers.Dropout (0.5, name= 'dropout_layer4')(layer_4)


# al final definimos la capa de salida                        
outputs = layers.Dense (1, name = 'output_layer') (layer_4) 

# y creamos el modelo 
model = keras.Model(inputs=inputs, outputs=outputs, name='my_model')

In [ ]:
# Compilación del modelo
model.compile(optimizer='adam', 
                  loss='mse',
                  metrics=['mae'])

In [ ]:
# No modifique el código
model.fit(x_train,
          y_train,
          epochs=200,
          batch_size=batch_size,
          validation_split=0.2,
          verbose=1)

Epoch 1/200
21/21 [==============================] - 1s 12ms/step - loss: 551.6513 - mae: 21.6653 - val_loss: 594.7106 - val_mae: 22.5204
Epoch 2/200
21/21 [==============================] - 0s 6ms/step - loss: 476.3293 - mae: 19.8707 - val_loss: 427.8578 - val_mae: 18.4932
Epoch 3/200
21/21 [==============================] - 0s 4ms/step - loss: 271.6260 - mae: 13.6000 - val_loss: 129.4023 - val_mae: 8.4934
Epoch 4/200
21/21 [==============================] - 0s 4ms/step - loss: 139.4461 - mae: 9.1579 - val_loss: 91.0836 - val_mae: 7.0030
Epoch 5/200
21/21 [==============================] - 0s 4ms/step - loss: 133.4792 - mae: 8.7761 - val_loss: 115.6657 - val_mae: 8.4266
Epoch 6/200
21/21 [==============================] - 0s 4ms/step - loss: 131.8579 - mae: 8.4037 - val_loss: 73.3221 - val_mae: 6.3369
Epoch 7/200
21/21 [==============================] - 0s 4ms/step - loss: 106.5650 - mae: 7.6672 - val_loss: 82.4343 - val_mae: 7.0102
Epoch 8/200
21/21 [==============================] -

In [ ]:
# No modifique el código
results = model.evaluate(x_test, y_test, verbose=1)
print('Test Loss: {}'.format(results))

4/4 [==============================] - 0s 3ms/step - loss: 27.4320 - mae: 3.8757
Test Loss: [27.432024002075195, 3.8757386207580566]


**Solución optima:** 
Reduciendo el batch_size de 32 a 16 hamos consegido una reducción del test lost de 18.9684 a 14.6281

<a name='1.3'></a>
## Cuestión 3: Utilice el mismo modelo de la cuestión anterior pero añadiendo un callback de early stopping. Obtenga un test loss inferior al del modelo anterior

In [ ]:
# definir el early stopping callback
es_callback = keras.callbacks.EarlyStopping(
 monitor ='val_loss',
 patience=15,
 verbose=1)

In [ ]:

model = tf.keras.models.Sequential()
inputs = keras.Input(shape=(13,), name='input_layer')

layer_1 = layers.Dense (64, activation='relu', name = 'layer_1') (inputs)
layer_2 = layers.Dense (64, activation='relu', name = 'layer_2') (layer_1)
layer_3 = layers.Dense (64, activation='relu', name = 'layer_3') (layer_2)
layer_4 = layers.Dense (64, activation='relu', name = 'layer_4') (layer_3)
                       
outputs = layers.Dense (1, name = 'output_layer') (layer_4) 

model_early_stopping = keras.Model(inputs=inputs, outputs=outputs, name='my_model') 

In [ ]:
# Compilación del modelo
model_early_stopping.compile(optimizer='adam', 
                  loss='mse',
                  metrics=['mae'])

In [ ]:
history_early_stopping = model_early_stopping.fit(
          x_train,
          y_train,
          epochs=200,
          batch_size=16,
          validation_split=0.2,
          verbose=1,
          callbacks=[es_callback]) 

Epoch 1/200
21/21 [==============================] - 1s 12ms/step - loss: 6.2269 - mae: 1.9589 - val_loss: 17.0569 - val_mae: 2.8717
Epoch 2/200
21/21 [==============================] - 0s 4ms/step - loss: 8.0053 - mae: 2.1154 - val_loss: 15.5717 - val_mae: 2.9568
Epoch 3/200
21/21 [==============================] - 0s 4ms/step - loss: 6.6877 - mae: 1.7764 - val_loss: 15.7925 - val_mae: 2.7386
Epoch 4/200
21/21 [==============================] - 0s 4ms/step - loss: 6.2121 - mae: 1.8526 - val_loss: 14.1410 - val_mae: 2.6551
Epoch 5/200
21/21 [==============================] - 0s 4ms/step - loss: 4.6054 - mae: 1.5279 - val_loss: 14.3588 - val_mae: 2.5753
Epoch 6/200
21/21 [==============================] - 0s 4ms/step - loss: 5.8015 - mae: 1.7084 - val_loss: 14.6579 - val_mae: 2.6517
Epoch 7/200
21/21 [==============================] - 0s 4ms/step - loss: 5.1998 - mae: 1.6925 - val_loss: 15.9076 - val_mae: 2.6843
Epoch 8/200
21/21 [==============================] - 0s 4ms/step - loss: 4.

In [ ]:
# No modifique el código
results = model_early_stopping.evaluate(x_test, y_test, verbose=1)
print('Test Loss: {}'.format(results))

4/4 [==============================] - 0s 3ms/step - loss: 16.0201 - mae: 2.7015
Test Loss: [16.020097732543945, 2.701547384262085]


<a name='1.4'></a>
## Cuestión 4: ¿Podría haberse usado otra función de activación de la neurona de salida? En caso afirmativo especifíquela.

La función de activación que hemos utilizado es la linea, que es la que por defecto se aplica cuando no se especifica. Podríamos haber utilizado una relu porque estamos se trata de una regresión positiva, es decir, que los resultados son positivos. 

<a name='1.5'></a>
## Cuestión 5:  ¿Qué es lo que una neurona calcula?

**a)** Una función de activación seguida de una suma ponderada  de las entradas.

**b)** Una suma ponderada  de las entradas seguida de una función de activación.

**c)** Una función de pérdida, definida sobre el target.

**d)** Ninguna  de las anteriores es correcta


b) Una suma ponderada de las entradas seguida de una función de activación

<a name='1.6'></a>
## Cuestión 6:  ¿Cuál de estas funciones de activación no debería usarse en una capa oculta (hidden layer)?

**a)** `sigmoid`

**b)** `tanh`

**c)** `relu`

**d)** `linear`


d) linear

<a name='1.7'></a>
## Cuestión 7:  ¿Cuál de estas técnicas es efectiva para combatir el overfitting en una red con varias capas ocultas? Ponga todas las que lo sean.

**a)** Dropout

**b)** Regularización L2.

**c)** Aumentar el tamaño del test set.

**d)** Aumentar el tamaño del validation set.

**e)** Reducir el número de capas de la red.

**f)** Data augmentation.

a) Dropout, b) Regularización L2, e) Reducir el número de capas de la red, f) Data augmentation

<a name='1.8'></a>
## Cuestión 8:  Supongamos que queremos entrenar una red para un problema de clasificación de imágenes con las siguientes clases: {'perro','gato','persona'}. ¿Cuántas neuronas y que función de activación debería tener la capa de salida? ¿Qué función de pérdida (loss function) debería usarse?


Un problema de clasificación en el que tenemos varias clases requiere de la la aplicación de una función de activación debe ser la softmax. 

La función de de pérdida adecuada es la categorical cross-entropy

<a name='actividad_2'></a>
# Actividad 2: Redes Convolucionales

Vamos a usar el dataset [cifar-10](https://www.cs.toronto.edu/~kriz/cifar.html), que son 60000 imágenes de 32x32 a color  con 10 clases diferentes. Para realizar mejor la práctica puede consultar [Introduction_to_CNN.ipynb](https://github.com/ezponda/intro_deep_learning/blob/main/class/CNN/Introduction_to_CNN.ipynb).



**Puntuación**: 

- [Cuestión 1](#2.1): 1 pt
- [Cuestión 2](#2.2): 1.5 pt
- [Cuestión 3](#2.3): 0.5 pts
- [Cuestión 4](#2.4): 0.5 pts
- [Cuestión 5](#2.5): 0.5 pts
- [Cuestión 6](#2.6): 0.5 pts




Puede normalizar las imágenes al principio o usar la capa [Rescaling](https://www.tensorflow.org/api_docs/python/tf/keras/layers/experimental/preprocessing/Rescaling):

```python
tf.keras.layers.experimental.preprocessing.Rescaling(
    scale, offset=0.0, name=None, **kwargs
)
```

In [ ]:
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.cifar10.load_data()
y_train = y_train.flatten()
y_test = y_test.flatten()

In [ ]:
class_names = ['airplane', 'automobile', 'bird', 'cat', 'deer',
               'dog', 'frog', 'horse', 'ship', 'truck']

plt.figure(figsize=(10,10))
for i in range(25):
    plt.subplot(5,5,i+1)
    plt.xticks([])
    plt.yticks([])
    plt.grid(False)
    plt.imshow(x_train[i])
    plt.xlabel(class_names[y_train[i]])
plt.show()

In [ ]:
print('x_train, y_train shapes:', x_train.shape, y_train.shape)
print('x_test, y_test shapes:', x_test.shape, y_test.shape)

<a name='2.1'></a>
## Cuestión 1: Cree una red convolucional con la API funcional con al menos dos capas convolucionales y al menos dos capas de pooling. Utilize sólo [Average Pooling](https://www.tensorflow.org/api_docs/python/tf/keras/layers/AveragePooling2D) y no añada ninguna regularización.

In [ ]:
inputs = tf.keras.Input(shape=..., name='input')
# reescaling = ...

# Convolution + pooling layers
...

# Flattening
...

# Fully-connected
outputs = layers.Dense(...)

model = keras.Model(inputs=..., outputs=...)

In [ ]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(),
              metrics=['accuracy'])

In [ ]:
history = model.fit(x_train, y_train, epochs=25, batch_size=64,
                    validation_split=0.15)

In [ ]:
results = model.evaluate(x_test, y_test, verbose=0, batch_size=1000)
print('Test Loss: {}'.format(results[0]))
print('Test Accuracy: {}'.format(results[1]))

<a name='2.2'></a>
## Cuestión 2: Cree un modelo con la API funcional con un máximo de 2 capas convolucionales y un máximo de 2 capas de pooling. Utilize  [Max Pooling](https://www.tensorflow.org/api_docs/python/tf/keras/layers/MaxPool2D) o [Average Pooling](https://www.tensorflow.org/api_docs/python/tf/keras/layers/AveragePooling2D) y  añada la regularización que quiera. Debe obtener un `Test accuracy > 0.68`

In [ ]:
inputs = tf.keras.Input(shape=..., name='input')
# reescaling = ...

# Convolution + pooling layers
...

# Flattening
...

# Fully-connected
outputs = layers.Dense(...)

model = keras.Model(inputs=..., outputs=...)

In [ ]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(),
              metrics=['accuracy'])

In [ ]:
history = model.fit(x_train, y_train, epochs=..., batch_size=...,
                    validation_split=0.15, callbacks=lbacks=[...])

In [ ]:
results = model.evaluate(x_test, y_test, verbose=0, batch_size=1000)
print('Test Loss: {}'.format(results[0]))
print('Test Accuracy: {}'.format(results[1]))

<a name='2.3'></a>
## Cuestión 3: Añada data augmentation al principio del modelo



In [ ]:
data_augmentation=... 

In [ ]:
inputs = tf.keras.Input(shape=..., name='input')
data_aug= ...

# reescaling = ...

# Convolution + pooling layers
...

# Flattening
...

# Fully-connected
outputs = layers.Dense(...)

model = keras.Model(inputs=..., outputs=...)

In [ ]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(),
              metrics=['accuracy'])

In [ ]:
history = model.fit(x_train, y_train, epochs=..., batch_size=...,
                    validation_split=0.15, callbacks=lbacks=[...])

In [ ]:
results = model.evaluate(x_test, y_test, verbose=0, batch_size=1000)
print('Test Loss: {}'.format(results[0]))
print('Test Accuracy: {}'.format(results[1]))

<a name='2.4'></a>
## Cuestión 4: Cree el mismo  modelo de manera secuencial. No es necesario compilar ni entrenar el modelo

In [ ]:
model_seq = tf.keras.models.Sequential()
# Código aquí
...

<a name='2.5'></a>
## Cuestión 5: Si tenenemos una  una imagen de entrada de 300 x 300 a color (RGB) y queremos usar una red densa. Si la primera capa oculta tiene 100 neuronas, ¿Cuántos parámetros tendrá esa capa (sin incluir el bias) ?


<a name='2.6'></a>
## Cuestión 6   Ponga  las verdaderas ventajas de las redes convolucionales respecto a las densas

**a)** Reducen el número total de parámetros, reduciendo así el overfitting.

**b)** Permiten utilizar una misma 'función'  en varias localizaciones de la imagen de entrada, en lugar de aprender una función diferente para cada pixel.

**c)** Permiten el uso del transfer learning.

**d)** Generalmente son menos profundas, lo que facilita su entrenamiento.



Las 4 son ventajas de las redes convolucionales respecto a las densas:

a) Reducen el número total de parámetros, reduciendo así el overfitting

b) Permiten utilizar una misma "función" en varias localizaciones de la imagen de entrada, en lugar de aprender una función diferente para cada pixel

c) Permite el uso del transfer learning

d) Generalmente son menos profundas, lo que facilita su entrenamiento